lib/services/database/post_upvote_service.dart

```diff
@@ -0,0 +1,131 @@
+import "package:cloud_firestore/cloud_firestore.dart";
+import "package:hooks_riverpod/hooks_riverpod.dart";
+import "package:proxima/models/database/post/post_data.dart";
+import "package:proxima/models/database/post/post_firestore.dart";
+import "package:proxima/models/database/post/post_id_firestore.dart";
+import "package:proxima/models/database/upvote_state.dart";
+import "package:proxima/models/database/user/user_id_firestore.dart";
+import "package:proxima/services/database/firestore_service.dart";
+
+/// This repository service is responsible for handling the upvotes of posts
+class PostUpvoteRepositoryService {
+  /// The name of the subcollection that contains the list of users who upvoted
+  static const upvotersSubCollectionName = "upvoters";
+
+  /// The name of the subcollection that contains the list of users who downvoted
+  static const downvotersSubCollectionName = "downvoters";
+
+  final FirebaseFirestore firestore;
+
+  PostUpvoteRepositoryService({
+    required this.firestore,
+  });
+
+  /// Returns the document reference of the post with id [postId]
+  DocumentReference<Map<String, dynamic>> _postDocument(
+    PostIdFirestore postId,
+  ) {
+    return firestore.collection(PostFirestore.collectionName).doc(postId.value);
+  }
+
+  /// Returns the collection reference of the subcollection that contains the
+  /// list of users who upvoted the post with id [postId]
+  CollectionReference<Map<String, dynamic>> _upvotersCollection(
+    PostIdFirestore postId,
+  ) {
+    return _postDocument(postId).collection(upvotersSubCollectionName);
+  }
+
+  /// Returns the collection reference of the subcollection that contains the
+  /// list of users who downvoted the post with id [postId]
+  CollectionReference<Map<String, dynamic>> _downvotersCollection(
+    PostIdFirestore postId,
+  ) {
+    return _postDocument(postId).collection(downvotersSubCollectionName);
+  }
+
+  /// Returns the upvote state of the user with id [userId] on the post with id [postId]
+  /// This is done atomically.
+  Future<UpvoteState> getUpvoteState(
```

Comment:
This will require modifying our current viemmodel for displaying a post, as for the moment we only have access to the displaying of the upvote count but nothing was made to accommodate the state of the upvote for a given post, for the current logged-in user.

I am not yet sure if we should create a new separate `PostUpVoteViewModel` or have it as a new attribute inside the `PostOverview`.
It is not really the goal of this PR anyway, but it is definitely something to keep in mind for future integration.

Reply:
I'm guessing this will go in the viewmodel of the overview, but we will indeed see later! :)